src/core/remio.c

```diff
@@ -0,0 +1,668 @@
+}
+
+/**
+ * @brief Push a new Remote I/O request event into the list of pending I/O requests events
+ * @param device Pointer to the Remote I/O device
+ * @param event Pointer to the Remote I/O request event
+ */
+static void remio_push_request_event(struct remio_device* device, struct remio_request_event* event)
+{
+    list_push(&device->request_event_list, (node_t*)event);
+}
+
+/**
+ * @brief Pop the next I/O request event from the list of pending I/O requests events
+ * @param device Pointer to the Remote I/O device
+ * @return Returns the next pending I/O request event or NULL if there is no pending I/O request
+ */
+static struct remio_request_event* remio_pop_request_event(struct remio_device* device)
+{
+    struct remio_request_event* event =
+        (struct remio_request_event*)list_pop(&device->request_event_list);
+    return event;
+}
+
+/**
+ * @brief Gets the number of pending I/O requests
+ * @param device Pointer to the Remote I/O device
+ * @return Returns the number of pending I/O requests
+ */
+static size_t remio_get_request_event_count(struct remio_device* device)
+{
+    return list_size(&device->request_event_list);
+}
+
+void remio_init(void)
+{
+    size_t frontend_cnt = 0, backend_cnt = 0;
+    int devices[REMIO_MAX_DEVICES][REMIO_NUM_DEV_TYPES];
+
+    /** Only execute the Remote I/O initialization routine on the master CPU */
+    if (!cpu_is_master()) {
+        return;
+    }
+
+    objpool_init(&remio_device_pool);
+    objpool_init(&remio_request_event_pool);
+    list_init(&remio_device_list);
+
+    for (size_t i = 0; i < REMIO_MAX_DEVICES; i++) {
+        devices[i][REMIO_DEV_FRONTEND] = REMIO_DEVICE_UNINITIALIZED;
+        devices[i][REMIO_DEV_BACKEND] = REMIO_DEVICE_UNINITIALIZED;
+    }
+
+    /** Create the Remote I/O devices based on the VM configuration */
+    for (size_t vm_id = 0; vm_id < config.vmlist_size; vm_id++) {
+        struct vm_config* vm_config = &config.vmlist[vm_id];
+        for (size_t i = 0; i < vm_config->platform.remio_dev_num; i++) {
+            struct remio_dev* dev = &vm_config->platform.remio_devs[i];
+            if (devices[dev->id][dev->type] != REMIO_DEVICE_UNINITIALIZED) {
+                ERROR("Failed to link backend to the frontend, more than one %s was "
+                      "atributed to the Remote I/O device %d",
+                    dev->type == REMIO_DEV_BACKEND ? "backend" : "frontend", dev->id);
+            }
+            if (dev->type == REMIO_DEV_BACKEND) {
+                struct remio_device* device = objpool_alloc(&remio_device_pool);
+                if (device == NULL) {
+                    ERROR("Failed allocating Remote I/O device node");
+                }
+                device->id = dev->id;
+                list_init(&device->request_event_list);
+                list_push(&remio_device_list, (node_t*)device);
+                backend_cnt++;
+                devices[dev->id][REMIO_DEV_BACKEND] = (int)vm_id;
+            } else {
```

Comment:
Test for `REMIO_DEV_FRONTEND` explicitly
There are other instances of this, please update them as well

Reply:
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/b0212dc3c64a3f834cf0c44a92d75b6f0ae354a4